libs/server/Auth/GarnetACLAuthenticator.cs

```diff
@@ -66,6 +70,22 @@ public bool Authenticate(ReadOnlySpan<byte> password, ReadOnlySpan<byte> usernam
                 string uname = Encoding.ASCII.GetString(username);
                 User user = string.IsNullOrEmpty(uname) ? _acl.GetDefaultUser() : _acl.GetUser(uname);
 
+                // Use injected authenticator if configured.
+                if (_authenticator != null)
```

Comment:
I am not very familiar with this part of the code but I think that this part does not look right? 
Aren't we supposed to override Authenticate based on the implementation of the authentication that is defined at startup?
Also, have you seen the existing implementation of GarnetAadAuthenticator? Maybe we need an extension of that implementation, instead of injecting it here?

Reply:
So the way I see it - we're trying to extend authz with another form of authn rather than other way around.
I have added another option for this mode AclWithAad which configures the ACLAuthenticator with an IGarnetAuthenticator and short circuits the response if such a Authenticator exists. 
Currently, the codebase has checks like ``` if (_authenticator is GarnetACLAuthenticator aclAuthenticator)``` and  ```(_authenticator.GetType() != typeof(GarnetACLAuthenticator))```
casting logic making it dependent on the GarnetACLAuthenticator, adding a GarnetACLWithAADAuthenticator would break a lot of things. There's ACL refresh logic also as part of this which would need to be ported to other classes. So above challenges made a strong case to reuse existing GarnetACLAuthenticator.

Now we can inject a default multi pass authenticator to cover default logic of scanning multiple passwords behind an IGarnetAuthenticator too but again on ACL refresh, this set will change. As a result, I stuck with default implementation with a short circuit if an IGarnetAuthenticator is defined. Incase of ACLWithAAD we're validating we're not relying on ACL password so don't need a refresh on ACL as well (only users which are scoped to ACLAuthenticator). An alternate approach I can think of is that we use an IACLAuthenticator instance which exposes two methods ```Authenticate(username, password)  and Refresh(ACLList)``` to with implementations being AclPasswordAuthenticator and AclAADAuthenticator. ACLPasswordAuthenticator can have default logic of looking at all passwords whereas in AclAADAuthenticator we can keep this current logic.

Reply:
The confusion is that you are treating GarnetACLAuthenticator as a GarnetAuthenticatorManager. 
I think it is a valid point that you want to use existing implementation, that is why I suggested creating an extension of GarnetACLAuthenticator.

``` csharp
    class GarnetACLWithAadAuthenticator : GarnetACLAuthenticator
    {
        public GarnetACLWithAadAuthenticator(AccessControlList accessControlList, ILogger logger) : base(accessControlList, logger)
        {

        }
```
It seems messy and hard to follow when interjecting another authenticator inside GarnetACLAuthenticator.
Your suggestion makes sense and I believe you don't need an IACLAuthenticator. You can add Refresh to IGarnetAuthenticator and just make it so that it throws unimplemented exception for any case that it is not needed.
I know it may be getting out of scope for the PR but I will let others chime in on how to proceed.
My vote is to address it now.